node/subsystem-util/src/validator_discovery.rs

```diff
@@ -104,6 +105,75 @@ async fn connect_to_authorities<Context: SubsystemContext>(
 	Ok((connected_rx, revoke_tx))
 }
 
+/// A struct that assists performing multiple concurrent connection requests.
+///
+/// This allows concurrent connections to validator sets at different `relay_parents`
+/// and multiplexes their results into a single `Stream`.
+#[derive(Default)]
+pub struct ConnectionRequests {
+	// added connection requests relay_parent -> StreamUnordered token
+	id_map: HashMap<Hash, usize>,
+
+	// Connection requests themselves.
+	requests: StreamUnordered<ConnectionRequest>,
+}
+
+impl ConnectionRequests {
+	/// Insert a new connection request.
+	///
+	/// If a `ConnectionRequest` under a given `relay_parent` already exists it will
+	/// be revoked and substituted with a new one.
+	pub fn put(&mut self, relay_parent: Hash, request: ConnectionRequest) {
+		self.remove(&relay_parent);
+		let token = self.requests.push(request);
+
+		self.id_map.insert(relay_parent, token);
+	}
+
+	/// Remove a connection request by a given `relay_parent`.
+	pub fn remove(&mut self, relay_parent: &Hash) {
+		if let Some(token) = self.id_map.remove(relay_parent) {
+			Pin::new(&mut self.requests).remove(token);
+		}
+	}
+}
+
+impl stream::Stream for ConnectionRequests {
+	/// (relay_parent, validator_id, peer_id).
+	type Item = (Hash, ValidatorId, PeerId);
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut task::Context) -> Poll<Option<Self::Item>> {
+		loop {
+			match Pin::new(&mut self.requests).poll_next(cx) {
+				Poll::Ready(Some((yeild, token))) => {
+					match yeild {
+						StreamYield::Item(item) => {
+							let relay_parent = self.id_map.iter()
+								.find_map(|(relay_parent, &val)|
+									if val == token {
+										Some(relay_parent)
+									} else {
+										None
+									}
+								)
+								.expect("All requests' streams should be accounted for; qed");
```

Comment:
Can the connection request not just return this data? So that we don't need the expect here?

Reply:
This has to do with tokens of the `StreamsUnordered` housekeeping, not what the requests return; if the tokens are broken or are not accounted for properly the code is broken.

Reply:
Yeah I know why you are doing this here. You could also wrap the connection request. This is just more clean and faster.